action/protocol/staking/staking_statereader.go

```diff
@@ -41,28 +42,134 @@ func newCompositeStakingStateReader(liquidIndexer LiquidStakingIndexer, nativeIn
+	// get height arg
+	inputHeight, err := c.nativeSR.SR().Height()
+	if err != nil {
+		return nil, 0, err
+	}
+	rp := rolldpos.MustGetProtocol(protocol.MustGetRegistry(ctx))
+	epochStartHeight := rp.GetEpochHeight(rp.GetEpochNum(inputHeight))
+
+	var (
+		buckets      *iotextypes.VoteBucketList
+		height       uint64
+		bucketsBytes []byte
+	)
+	if epochStartHeight != 0 && c.nativeIndexer != nil {
+		// read native buckets from indexer
+		bucketsBytes, height, err = c.nativeIndexer.GetBuckets(epochStartHeight, req.GetPagination().GetOffset(), req.GetPagination().GetLimit())
+		if err != nil {
+			return nil, 0, err
+		}
+		buckets = &iotextypes.VoteBucketList{}
+		if err := proto.Unmarshal(bucketsBytes, buckets); err != nil {
+			return nil, 0, err
+		}
+	} else {
+		// read native buckets from state
+		buckets, height, err = c.nativeSR.readStateBuckets(ctx, req)
+		if err != nil {
+			return nil, 0, err
+		}
+	}
+	// read LSD buckets
+	lsdBuckets, err := c.liquidIndexer.Buckets()
+	if err != nil {
+		return nil, 0, err
+	}
+	lsdIoTeXBuckets, err := toIoTeXTypesVoteBucketList(lsdBuckets)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// merge native and LSD buckets
+	buckets.Buckets = append(buckets.Buckets, lsdIoTeXBuckets.Buckets...)
+	buckets.Buckets = getPageOfArray(buckets.Buckets, int(req.GetPagination().GetOffset()), int(req.GetPagination().GetLimit()))
+	return buckets, height, err
 }
 
 func (c *compositeStakingStateReader) readStateBucketsByVoter(ctx context.Context, req *iotexapi.ReadStakingDataRequest_VoteBucketsByVoter) (*iotextypes.VoteBucketList, uint64, error) {
-	// TODO (iip-13): combine native and liquid staking buckets
-	return c.nativeSR.readStateBucketsByVoter(ctx, req)
+	// read native buckets
+	buckets, height, err := c.nativeSR.readStateBucketsByVoter(ctx, req)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// read LSD buckets
+	lsdBuckets, err := c.liquidIndexer.Buckets()
+	if err != nil {
+		return nil, 0, err
+	}
+	lsdBuckets = filterBucketsByVoter(lsdBuckets, req.GetVoterAddress())
+	lsdIoTeXBuckets, err := toIoTeXTypesVoteBucketList(lsdBuckets)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// merge native and LSD buckets
+	buckets.Buckets = append(buckets.Buckets, lsdIoTeXBuckets.Buckets...)
+	buckets.Buckets = getPageOfArray(buckets.Buckets, int(req.GetPagination().GetOffset()), int(req.GetPagination().GetLimit()))
+	return buckets, height, err
 }
 
 func (c *compositeStakingStateReader) readStateBucketsByCandidate(ctx context.Context, req *iotexapi.ReadStakingDataRequest_VoteBucketsByCandidate) (*iotextypes.VoteBucketList, uint64, error) {
-	// TODO (iip-13): combine native and liquid staking buckets
-	return c.nativeSR.readStateBucketsByCandidate(ctx, req)
+	// read native buckets
+	buckets, height, err := c.nativeSR.readStateBucketsByCandidate(ctx, req)
+	if err != nil {
+		return nil, 0, err
+	}
+	// read LSD buckets
+	lsdBuckets, err := c.liquidIndexer.Buckets()
```

Comment:
indexer should be able to provide a `BucketsByCandidate()` API, since internally it has that `map[string]map[uint64]bool` we chatted today

Reply:
fixed